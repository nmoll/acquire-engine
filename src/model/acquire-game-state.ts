export type AcquireGameState = "not started" | "started";
