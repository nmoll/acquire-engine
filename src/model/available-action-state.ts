import { AvailableAction } from "./available-action";

export type IAvailableActionState = AvailableAction[];
