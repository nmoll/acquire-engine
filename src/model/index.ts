export * from "./board-square-state";
export * from "./game-state";
export * from "./hotel-chain-type";
export * from "./player-turn";
export * from "./shares-state";
