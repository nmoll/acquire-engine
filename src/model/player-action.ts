import { HotelChainType } from "./hotel-chain-type";
import { IShares } from "./shares";

export interface PlaceTile {
  type: "PlaceTile";
  playerId: number;
  boardSquareId: number;
}

export interface StartHotelChain {
  type: "StartHotelChain";
  playerId: number;
  hotelChain: HotelChainType;
}

export interface PurchaseShares {
  type: "PurchaseShares";
  playerId: number;
  shares: IShares[];
}

export interface ChooseMergeDirection {
  type: "ChooseMergeDirection";
  playerId: number;
  hotelChainToKeep: HotelChainType;
}

export type PlayerAction =
  | PlaceTile
  | StartHotelChain
  | PurchaseShares
  | ChooseMergeDirection;

export const PlayerActionType = {
  PlaceTile: (playerId: number, boardSquareId: number): PlaceTile => ({
    type: "PlaceTile",
    playerId,
    boardSquareId,
  }),
  StartHotelChain: (
    playerId: number,
    hotelChain: HotelChainType
  ): StartHotelChain => ({
    type: "StartHotelChain",
    playerId,
    hotelChain,
  }),

  PurchaseShares: (playerId: number, shares: IShares[]): PurchaseShares => ({
    type: "PurchaseShares",
    playerId,
    shares,
  }),

  ChooseMergeDirection: (
    playerId: number,
    hotelChainToKeep: HotelChainType
  ): ChooseMergeDirection => ({
    type: "ChooseMergeDirection",
    playerId,
    hotelChainToKeep,
  }),
};
