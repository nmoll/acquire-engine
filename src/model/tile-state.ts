export interface ITileState {
  availableTiles: number[];
  [playerId: number]: number[];
}
