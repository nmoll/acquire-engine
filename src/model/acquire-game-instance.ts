import { AcquireGameState } from "./acquire-game-state";

export interface IAcquireGameInstance {
  id: string;
  randomSeed: number;
  playerIds: string[];
  hostId: string;
  state: AcquireGameState;
  isOpen?: boolean;
  createdDate?: number;
}
