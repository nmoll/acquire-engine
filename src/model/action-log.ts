import { IGameState } from "./game-state";
import { PlayerAction } from "./player-action";
import { PlayerActionResult } from "./player-action-result";

export interface ActionLog {
  action: PlayerAction;
  actionResult: PlayerActionResult;
  state: IGameState;
}
