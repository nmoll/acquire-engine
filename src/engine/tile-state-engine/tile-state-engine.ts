import { GameConfig } from "../../game-config";
import { IAcquireGameInstance } from "../../model/acquire-game-instance";
import { PlayerActionResult } from "../../model/player-action-result";
import { ITileState } from "../../model/tile-state";
import { TileUtils } from "../../utils/tile-utils";

const getInitialState = (gameInstance: IAcquireGameInstance): ITileState => {
  const tileBag = TileUtils.getSortedBag(gameInstance.randomSeed);

  return gameInstance.playerIds.reduce<ITileState>(
    (res, playerId) =>
      addPlayerTiles(GameConfig.tile.rackSize, playerId, tileBag, res),
    {}
  );
};

export const computeState = (
  gameInstance: IAcquireGameInstance,
  actionResult: PlayerActionResult,
  tileState: ITileState
): ITileState => {
  const tileBag = TileUtils.getSortedBag(gameInstance.randomSeed);
  const playerAction = actionResult.action;

  if (playerAction.type === "PlaceTile") {
    return discardPlayerTile(
      playerAction.playerId,
      playerAction.boardSquareId,
      tileState
    );
  } else if (playerAction.type === "EndTurn") {
    return addPlayerTile(playerAction.playerId, tileBag, tileState);
  } else {
    return tileState;
  }
};

const addPlayerTile = (
  playerId: string,
  tileBag: number[],
  tileState: ITileState
): ITileState => {
  const nextTile = TileUtils.getNextTile(tileBag, tileState);

  if (nextTile === undefined) {
    return tileState;
  }

  return {
    ...tileState,
    [playerId]: [...(tileState[playerId] ?? []), nextTile],
  };
};

const addPlayerTiles = (
  amount: number,
  playerId: string,
  tileBag: number[],
  tileState: ITileState
): ITileState => {
  if (amount >= 1) {
    const stateAfterAdd = addPlayerTile(playerId, tileBag, tileState);
    return addPlayerTiles(amount - 1, playerId, tileBag, stateAfterAdd);
  } else {
    return tileState;
  }
};

const discardPlayerTile = (
  playerId: string,
  tileId: number,
  tileState: ITileState
): ITileState => ({
  ...tileState,
  [playerId]: tileState[playerId].filter(
    (playerTileId) => playerTileId !== tileId
  ),
});

export const TileStateEngine = {
  getInitialState,
  computeState,
};
