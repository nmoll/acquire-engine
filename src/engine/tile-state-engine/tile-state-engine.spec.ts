import { IAcquireGameInstance } from "../../model/acquire-game-instance";
import { PlayerAction } from "../../model/player-action";
import { ITileState } from "../../model/tile-state";
import { TileStateEngine } from "./tile-state-engine";

describe("TileStateEngine", () => {
  let gameInstance: IAcquireGameInstance;

  beforeEach(() => {
    gameInstance = {
      randomSeed: 1,
      playerIds: [1, 2, 3, 4],
    };
  });

  it("should return initial tile state if no given state", () => {
    expect(TileStateEngine.computeState(gameInstance)).toMatchSnapshot();
  });

  it("should return given tile state if no action", () => {
    const tileState: ITileState = {
      1: [5, 2, 8, 7],
    };

    expect(TileStateEngine.computeState(gameInstance, null, tileState)).toEqual(
      tileState
    );
  });

  it("should return the same initial state for the same game instance", () => {
    expect(TileStateEngine.computeState(gameInstance)).toEqual(
      TileStateEngine.computeState(gameInstance)
    );
  });

  it("should remove player's tile when played", () => {
    const tileState: ITileState = {
      1: [4, 7, 21, 34, 19],
      2: [15, 84, 100, 3, 1],
      3: [2, 8, 11, 13, 62],
    };
    const tilePlaced = 7;
    const action: PlayerAction = {
      type: "PlaceTile",
      playerId: 1,
      boardSquareId: tilePlaced,
    };

    const expected: ITileState = {
      1: [4, 21, 34, 19],
      2: [15, 84, 100, 3, 1],
      3: [2, 8, 11, 13, 62],
    };

    expect(
      TileStateEngine.computeState(gameInstance, action, tileState)
    ).toEqual(expected);
  });

  it("should give player next tile when turn ends", () => {
    const initialState = TileStateEngine.computeState(gameInstance);
    const action: PlayerAction = {
      type: "EndTurn",
      playerId: 1,
    };

    const resultState = TileStateEngine.computeState(
      gameInstance,
      action,
      initialState
    );

    expect(resultState[1]).toEqual([11, 7, 63, 76, 46, 30]);
  });

  it("should do nothing if there are no tiles left", () => {
    const tileState: ITileState = {
      // The last tiles in the bag for this game seed
      1: [0, 67],
    };
    const action: PlayerAction = {
      type: "EndTurn",
      playerId: 1,
    };

    const resultState = TileStateEngine.computeState(
      gameInstance,
      action,
      tileState
    );

    expect(resultState).toEqual(tileState);
  });
});
