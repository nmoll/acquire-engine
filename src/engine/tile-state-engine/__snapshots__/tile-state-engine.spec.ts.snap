// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileStateEngine should return initial tile state if no given state 1`] = `
Object {
  "1": Array [
    11,
    7,
    63,
    76,
    46,
  ],
  "2": Array [
    107,
    15,
    71,
    75,
    50,
  ],
  "3": Array [
    106,
    29,
    81,
    79,
    101,
  ],
  "4": Array [
    65,
    60,
    32,
    70,
    21,
  ],
}
`;
